### PR TITLE
Extend component types

### PR DESCRIPTION
**Pull request by atwixfirster**
## Purpose of this pull request

This pull request (PR) adds `setup` component type.

## Affected DevDocs pages

- https://devdocs.magento.com/guides/v2.3/extension-dev-guide/prepare/dev-modtypes.html

## Links to Magento source code

- [ComponentRegistrar.php](https://github.com/magento/magento2/blob/2.3/lib/internal/Magento/Framework/Component/ComponentRegistrar.php#L25)

Thank you!

### CODE REVIEW

**Comment by devops-devdocs** [comment]
An admin must run tests on this PR before it can be merged.

**Comment by atwixfirster** [comment]
@jeff-matthews 
> We need a better description of "Setup" in the table.

Agreed but unfortunately I did not work with "Setup" component. Also I did not find an example of usage in Magento CE. This section may be updated later.

What do you think?

Thank you!

**Comment by jeff-matthews** [comment]
I did a little more searching and found a description of a module "setup" directory:

> Setup: contains classes for module database structure and data setup which are invoked when installing or upgrading.

Is it accurate to describe a "setup module" the same way?  

https://devdocs.magento.com/guides/v2.3/extension-dev-guide/build/module-file-structure.html#common-directories

**Comment by atwixfirster** [comment]
> Is it accurate to describe a "setup module" the same way?

 @jeff-matthews, unfortunately, no...

We are talking about COMPONENT TYPE (about module type) here. In the mentioned by you link - the talk is about module folders structure.

Thanks

**Comment by jeff-matthews** [comment]
Thanks @atwixfirster. I'll check with an internal developer to get a better description.

**Comment by atwixfirster** [comment]
> Thanks @atwixfirster. I'll check with an internal developer to get a better description.

Cool. We need someone who developed this type of component. This will a very useful information for community.

Thank you, @jeff-matthews ! 👍

**Comment by jeff-matthews** [comment]
@buskamuza, can you help provide an accurate description of the `setup` component type or let us know who can?

**Comment by atwixfirster** [comment]
@buskamuza, could you please help us here?

Many thanks! 🌸 

CC: @jeff-matthews

**Comment by buskamuza** [comment]
Most types are described in the Composer plugin https://github.com/magento/magento-composer-installer that uses these types to deploy the package correctly.
There might be other types, which are not used by this plugin and so I don't know where to find their description. For now, I'm not aware of any such types existing. MFTF team is adding a type for functional tests currently, they should sync with DevDocs team to add the new type to this page.

To be honest, I'm not aware of `setup` type. If this is about `Setup` folder, it is part of `magento2-base` package, which includes all the files that do not belong to any module/theme/etc, but still can't live just inside `vendor` folder, it must be copied to the Magento root folder. But this package uses `magento2-component`, so the Composer plugin can copy the files in the root.

I'd suggest to start the page from the documentation of the plugin I mentioned. And then sync with @okolesnyk on what types functional tests may add.

**Comment by atwixfirster** [comment]
@jeff-matthews , I 've skipped info about setup component regarding [Magento Composer Installer](https://github.com/magento/magento-composer-installer/blob/master/README.md) documentation.

@rogyar , could you please re-check this PR and let me know if you have any notices here?

Thank you!

File: _includes/php-dev/composer-types.md [hunk 1]
```diff
@@ -6,4 +6,5 @@ The following table discusses the component types that Magento Marketplace suppo
 |Module|magento2-module|Code that modifies Magento application behavior. You can upload a single module to the Magento Marketplace or your module can be dependent on some parent package.|
 |Theme|magento2-theme|Code that modifies the look and feel of the storefront or Magento Admin.|
 |Language package|magento2-language|Translations for the storefront or Admin.|
-
+|Library|magento2-library|Support for libraries located in `lib/internal` instead of in the `vendor` directory.|
+|Component|magento2-component|Good example of `magento2-component` is the [Magento 2 Data Migration Tool]({{ site.mage2bloburl }}/{{ page.guide_version }}/migration/bk-migration-guide.html).|
```
Comment (buskamuza) [inline _includes/php-dev/composer-types.md]: Why is it the Migration Tool? It's not a `magento2-component` from what I can see - https://github.com/magento/data-migration-tool/blob/2.3-develop/composer.json
So far the only example is the Magento base package - the package formed of the files that must be located in root (`index.php`, `.htaccess`, etc). This includes `dev/tests` and `setup` as well for now.

Comment (atwixfirster) [reply _includes/php-dev/composer-types.md]: > Why is it the Migration Tool?

Regarding to https://github.com/magento/magento-composer-installer/blob/master/README.md M2 Migration Tool was developed as a magento-component :)

However, I've checked [MagentoHackathon\Composer\Magento\PackageTypes](https://www.magentoextensions.org/documentation/class_magento_hackathon_1_1_composer_1_1_magento_1_1_package_types.html) and found 
the files that must be located in root .

I've **fixed a provided** description according to your comment. Could you please review?

Thank you, @buskamuza !

Comment (buskamuza) [reply _includes/php-dev/composer-types.md]: I see... looks like it's something outdated.
All looks good now.

**Comment by jeff-matthews** [comment]
running tests

**Pull request merged**